url4c41254592_topskychen_AuthTopk_tgz-pJ8-quickhull3d_QuickHull3DTestJ8/src/spatialindex/io/DiskStorageManager.java

```diff
@@ -0,0 +1,423 @@
+		}
+
+		// find page size.
+		if (bOverwrite == true) {
+			var = ps.getProperty("PageSize");
+
+			if (var != null) {
+				if (!(var instanceof Integer)) {
+					throw new IllegalArgumentException(
+							"Property PageSize must be an Integer");
+				}
+				m_pageSize = ((Integer) var).intValue();
+				m_nextPage = 0;
+			} else {
+				throw new IllegalArgumentException(
+						"Property PageSize was not specified.");
+			}
+		} else {
+			try {
+				/*
+				 * This means page-size is stored at first
+				 */
+				m_pageSize = m_indexFile.readInt();
+			} catch (EOFException ex) {
+				throw new IllegalStateException("Failed reading pageSize.");
+			}
+
+			try {
+				/*
+				 * next-page is stored at 2nd
+				 */
+				m_nextPage = m_indexFile.readInt();
+			} catch (EOFException ex) {
+				throw new IllegalStateException("Failed reading nextPage.");
+			}
+		}
+
+		// create buffer.
+		m_buffer = new byte[m_pageSize];
+
+		/**
+		 * load exist data
+		 */
+		if (bOverwrite == false) {
+			int count, id, page;
+
+			// load empty pages in memory.
+			try {
+				count = m_indexFile.readInt();
+
+				for (int cCount = 0; cCount < count; cCount++) {
+					page = m_indexFile.readInt();
+					m_emptyPages.add(new Integer(page));
+				}
+
+				// load index table in memory.
+				count = m_indexFile.readInt();
+
+				for (int cCount = 0; cCount < count; cCount++) {
+					Entry e = new Entry();
+
+					id = m_indexFile.readInt();
+					e.m_length = m_indexFile.readInt();
+
+					int count2 = m_indexFile.readInt();
+
+					for (int cCount2 = 0; cCount2 < count2; cCount2++) {
+						page = m_indexFile.readInt();
+						e.m_pages.add(new Integer(page));
+					}
+					m_pageIndex.put(new Integer(id), e);
+				}
+			} catch (EOFException ex) {
+				throw new IllegalStateException("Corrupted index file.");
+			}
+		}
+	}
+
+	/**
+	 * save the index to file
+	 */
+	@Override
+	public void flush() {
+		try {
+			m_indexFile.seek(0l);
+
+			m_indexFile.writeInt(m_pageSize);
+			m_indexFile.writeInt(m_nextPage);
+
+			int id, page;
+			int count = m_emptyPages.size();
+
+			m_indexFile.writeInt(count);
+
+			Iterator<Integer> it = m_emptyPages.iterator();
+			while (it.hasNext()) {
+				page = it.next().intValue();
+				m_indexFile.writeInt(page);
+			}
+
+			count = m_pageIndex.size();
+			m_indexFile.writeInt(count);
+
+			Iterator<Map.Entry<Integer, DiskStorageManager.Entry>> it2 = m_pageIndex
+					.entrySet().iterator();
+
+			while (it2.hasNext()) {
+				Map.Entry<Integer, DiskStorageManager.Entry> me = it2.next();
+				id = me.getKey().intValue();
+				m_indexFile.writeInt(id);
+
+				Entry e = me.getValue();
+				count = e.m_length;
+				m_indexFile.writeInt(count);
+
+				count = e.m_pages.size();
+				m_indexFile.writeInt(count);
+
+				for (int cIndex = 0; cIndex < count; cIndex++) {
+					page = e.m_pages.get(cIndex).intValue();
+					m_indexFile.writeInt(page);
+				}
+			}
+		} catch (IOException ex) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)